### PR TITLE
Wence/patch c wrapper

### PR DESCRIPTION
**Pull request by wence-**
This change significantly improves the performance of snespatch by avoiding cross-calling from C into Python and back quite so often.

The only particularly ugly thing is the string pasting for the C code.

@pefarrell, @florianwechsung, can you confirm this doesn't break existing things?

### CODE REVIEW

**Comment by wence-** [comment]
I think at @tlroy has been using this, so perhaps he can review too :).

**Comment by tlroy** [comment]
The code looks fine to me. It works on SNESPatch test cases, with some substantial speed-up in many cases.

File: firedrake/dmplex.pyx [hunk 1]
```diff
@@ -2371,3 +2371,39 @@ def clear_adjacency_callback(PETSc.DM dm not None):
         dm.removeLabel("ghost_region")
         CHKERR(DMLabelDestroy(&label))
     CHKERR(DMPlexSetAdjacencyUser(dm.dm, NULL, NULL))
+
+
+def set_patch_residual(patch, function, ctx, is_snes=False, interior_facets=False):
```
Comment (pefarrell) [inline firedrake/dmplex.pyx]: Should this really live in dmplex.pyx? Alternatively, should dmplex.pyx be renamed?

File: firedrake/dmplexinc.pxi [hunk 2]
```diff
@@ -68,6 +68,34 @@ cdef extern from "petscsf.h" nogil:
     int PetscSFReduceBegin(PETSc.PetscSF,MPI.MPI_Datatype,const void*, void*,MPI.MPI_Op)
     int PetscSFReduceEnd(PETSc.PetscSF,MPI.MPI_Datatype,const void*, void*,MPI.MPI_Op)
 
+ctypedef int (*PetscPCPatchComputeFunction)(PETSc.PetscPC,
```
Comment (pefarrell) [inline firedrake/dmplexinc.pxi]: Ugly, but unavoidable, duplication of stuff from petsc4py.

File: firedrake/mesh.py [hunk 3]
```diff
@@ -594,9 +594,8 @@ def merge_ids(x, y, datatype):
                                    self._cell_numbering,
                                    self.cell_closure)
 
-        point2facetnumber = {}
-        for i, f in enumerate(facets):
-            point2facetnumber[f] = i
+        point2facetnumber = np.full(facets.max(initial=0)+1, -1, dtype=IntType)
```
Comment (pefarrell) [inline firedrake/mesh.py]: Can you explain these changes?

**Comment by pefarrell** [comment]
I confirm this leads to substantial speedups, and has worked for every problem I tried it on.

**Comment by florianwechsung** [comment]
Looks good to me, everything still runs fine

**Comment by wence-** [comment]
Addressed the review comments, let's hope I didn't muck up the refactoring.

**Comment by wence-** [comment]
>  let's hope I didn't muck up the refactoring.

Reader, I did.

**Pull request merged**